README
    _ _ _     _                         _ _
   | (_) |__ | | _____  ___ _ __   __ _| (_)_   _____
   | | | '_ \| |/ / _ \/ _ \ '_ \ / _` | | \ \ / / _ \
   | | | |_) |   <  __/  __/ |_) | (_| | | |\ V /  __/
   |_|_|_.__/|_|\_\___|\___| .__/ \__,_|_|_| \_/ \___|
                           |_|

   (C) Fabio Busatto <fabio.busatto@sikurezza.org>



1. OVERVIEW
   Many systems provide the ability to keep tcp connections alive, so they
   aren't reset by peers or by routers because of inactivity.

   Even if this feature is present, only a few programs correctly implement
   the code to use it. If your favourite foo-client is not written with
   support for tcp keepalive, you'll continue to see your connection reset.

   libkeepalive library provides a way to enable tcp keepalive support in
   any program that uses shared libraries (e.g.: glibc shared object) to
   perform network operations. Using the preload method, you will be able to
   intercept normal program execution and to inject the code needed to enable
   the keepalive routines, everything done without modifying the original
   binary file and with no need to gain root privileges.

2. OBTAINING SOFTWARE
   Latest version of libkeepalive is available on
   http://libkeepalive.sourceforge.net/#download.

3. BUILDING BINARIES
   Just run make (use gmake if your default make is another one).

4. INSTALLING
   Simply copy libkeepalive.so in some cool directory (/usr/lib, or ~/lib if
   you're not so elite to have root privileges).
   If you want to enable keepalive in suid programs too, you have to copy
   libkeepalive.so in /lib or /usr/lib and suid it (according to ld.so
   manpage). In any other scenarios the library will not work with suid
   binaries.

5. USER CONFIGURATION
   libkeepalive will do its job if you set the environment variable LD_PRELOAD
   to the path of libkeepalive.so shared library before executing the target
   program.  You must specify the full path of the library if ld.so cannot is
   not able to find it. You can also set the variable in your startup script
   (i.e.: ~/.profile), depending from your shell.
   For informations about setting environment variables, refer to your shell
   documentation.

6. SYSTEM-WIDE CONFIGURATION
   Host administrators may want to impose the use of libkeepalive to all their
   users and all daemons loaded by the system startup process.
   Adding the full path of the libkeepalive.so dynamic library to
   /etc/ld.so.preload you obtain the same result as the LD_PRELOAD variable is
   always set, forcing the use indipendently from the environment.

7. DISABLING FOR A SPECIFIC EXECUTION
   If you need to disable the action of libkeepalive for a single job, or if
   your admin set the ld.so.preload, you can set the KEEPALIVE environment
   variable to "off", inhibiting the library even if it will be loaded.

8. TUNING KEEPALIVE PARAMETERS
   You can change the keepalive parameters using environment variables or via
   sysctl/procfs.

       env             sysctl
   KEEPCNT     <=>   net.ipv4.tcp_keepalive_probes
   KEEPIDLE    <=>   net.ipv4.tcp_keepalive_time
   KEEPINTVL   <=>   net.ipv4.tcp_keepalive_intvl

DON'T FORGET to configure your system-specific parameters for keepalive!
Many times, defaults are not what you need and you have to change your
settings (see your system documentation for this).


ChangeLog
libkeepalive-0.2:
   added support for runtime tuning using environment variables
libkeepalive-0.1:
   released first version of libkeepalive


test/test.c
/*
    _ _ _     _                         _ _
   | (_) |__ | | _____  ___ _ __   __ _| (_)_   _____
   | | | '_ \| |/ / _ \/ _ \ '_ \ / _` | | \ \ / / _ \
   | | | |_) |   <  __/  __/ |_) | (_| | | |\ V /  __/
   |_|_|_.__/|_|\_\___|\___| .__/ \__,_|_|_| \_/ \___|
                           |_|
*/

/* Fabio Busatto */

/*
  This program is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation; either version 2 of the License, or
  (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program (see the file COPYING included with this
  distribution); if not, write to the Free Software Foundation, Inc.,
  59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/

#include <stdio.h>
#include <stdlib.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <netinet/tcp.h>

int main(void);

int main()
{
   int s;
   int optval;
   socklen_t optlen = sizeof(optval);

   if((s = socket(PF_INET, SOCK_STREAM, IPPROTO_TCP)) < 0) {
      perror("socket()");
      exit(EXIT_FAILURE);
   }

   if(getsockopt(s, SOL_SOCKET, SO_KEEPALIVE, &optval, &optlen) < 0) {
      perror("getsockopt()");
      close(s);
      exit(EXIT_FAILURE);
   }
   printf("SO_KEEPALIVE is %s\n", (optval ? "ON" : "OFF"));

   if(optval) {
#ifdef TCP_KEEPCNT
      if(getsockopt(s, SOL_TCP, TCP_KEEPCNT, &optval, &optlen) < 0) {
         perror("getsockopt()");
         close(s);
         exit(EXIT_FAILURE);
      }
      printf("TCP_KEEPCNT   = %d\n", optval);
#endif
   
#ifdef TCP_KEEPIDLE
      if(getsockopt(s, SOL_TCP, TCP_KEEPIDLE, &optval, &optlen) < 0) {
         perror("getsockopt()");
         close(s);
         exit(EXIT_FAILURE);
      }
      printf("TCP_KEEPIDLE  = %d\n", optval);
#endif
   
#ifdef TCP_KEEPINTVL
      if(getsockopt(s, SOL_TCP, TCP_KEEPINTVL, &optval, &optlen) < 0) {
         perror("getsockopt()");
         close(s);
         exit(EXIT_FAILURE);
      }
      printf("TCP_KEEPINTVL = %d\n", optval);
#endif
   }
   
   close(s);

   exit(EXIT_SUCCESS);
}


src/libkeepalive.c
/*
    _ _ _     _                         _ _
   | (_) |__ | | _____  ___ _ __   __ _| (_)_   _____
   | | | '_ \| |/ / _ \/ _ \ '_ \ / _` | | \ \ / / _ \
   | | | |_) |   <  __/  __/ |_) | (_| | | |\ V /  __/
   |_|_|_.__/|_|\_\___|\___| .__/ \__,_|_|_| \_/ \___|
                           |_|
*/

/* Fabio Busatto */

/*
  This program is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation; either version 2 of the License, or
  (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program (see the file COPYING included with this
  distribution); if not, write to the Free Software Foundation, Inc.,
  59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/

#ifndef RTLD_NEXT
#  define _GNU_SOURCE
#endif
#include <dlfcn.h>
#include <errno.h>
#include <stdlib.h>
#include <strings.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/tcp.h>

int socket(int domain, int type, int protocol);

int socket(int domain, int type, int protocol)
{
  int (*libc_socket)(int, int, int);
  int s, optval;
  char *env;

  *(void **)(&libc_socket) = dlsym(RTLD_NEXT, "socket");
  if(dlerror()) {
    errno = EACCES;
    return -1;
  }

  if((s = (*libc_socket)(domain, type, protocol)) != -1) {
    if((domain == PF_INET) && (type == SOCK_STREAM)) {
      if(!(env = getenv("KEEPALIVE")) || strcasecmp(env, "off")) {
        optval = 1;
      } else {
        optval = 0;
      }
      if(!(env = getenv("KEEPALIVE")) || strcasecmp(env, "skip")) {
        setsockopt(s, SOL_SOCKET, SO_KEEPALIVE, &optval, sizeof(optval));
      }
#ifdef TCP_KEEPCNT
      if((env = getenv("KEEPCNT")) && ((optval = atoi(env)) >= 0)) {
        setsockopt(s, SOL_TCP, TCP_KEEPCNT, &optval, sizeof(optval));
      }
#endif
#ifdef TCP_KEEPIDLE
      if((env = getenv("KEEPIDLE")) && ((optval = atoi(env)) >= 0)) {
        setsockopt(s, SOL_TCP, TCP_KEEPIDLE, &optval, sizeof(optval));
      }
#endif
#ifdef TCP_KEEPINTVL
      if((env = getenv("KEEPINTVL")) && ((optval = atoi(env)) >= 0)) {
        setsockopt(s, SOL_TCP, TCP_KEEPINTVL, &optval, sizeof(optval));
      }
#endif
    }
  }

   return s;
}
